Preprocessing Pipeline (See bottom for x,y code.)


1. Read original dataset.

In [57]:
#Read Original Dataset
#Graduate: 2209/4424 
#Dropout: 1421/4424
#Enrolled: 794/4424


import pandas as pd
df = pd.read_csv("data/dataset.csv")
df.head()
# count_row=df2.shape[0]
# print(count_row)

Marital status  Application mode  Application order  Course  \
0               1                 8                  5       2   
1               1                 6                  1      11   
2               1                 1                  5       5   
3               1                 8                  2      15   
4               2                12                  1       3   

   Daytime/evening attendance  Previous qualification  Nacionality  \
0                           1                       1            1   
1                           1                       1            1   
2                           1                       1            1   
3                           1                       1            1   
4                           0                       1            1   

   Mother's qualification  Father's qualification  Mother's occupation  ...  \
0                      13                      10                    6  ...   
1                       1                       3                    4  ...   
2                      22                      27                   10  ...   
3                      23                      27                    6  ...   
4                      22                      28                   10  ...   

   Curricular units 2nd sem (credited)  Curricular units 2nd sem (enrolled)  \
0                                    0                                    0   
1                                    0                                    6   
2                                    0                                    6   
3                                    0                                    6   
4                                    0                                    6   

   Curricular units 2nd sem (evaluations)  \
0                                       0   
1                                       6   
2                                       0   
3                                      10   
4                                       6   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                    0                          0.000000   
1                                    6                         13.666667   
2                                    0                          0.000000   
3                                    5                         12.400000   
4                                    6                         13.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                               0               10.8   
1                                               0               13.9   
2                                               0               10.8   
3                                               0                9.4   
4                                               0               13.9   

   Inflation rate   GDP    Target  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 35 columns]

2. Delete all targets 'enrolled' not relevant class for us.

In [58]:
#Remove lines with target 'enrolled'. As not necessary to our research question.
#Graduate: 2209/3630
#Dropout: 1421/3630

df2 =  df[df.Target != "Enrolled"]
df2 = df2.reset_index(drop=True)
df2.head() 
# count_row=df2.shape[0]
# print(count_row)

Marital status  Application mode  Application order  Course  \
0               1                 8                  5       2   
1               1                 6                  1      11   
2               1                 1                  5       5   
3               1                 8                  2      15   
4               2                12                  1       3   

   Daytime/evening attendance  Previous qualification  Nacionality  \
0                           1                       1            1   
1                           1                       1            1   
2                           1                       1            1   
3                           1                       1            1   
4                           0                       1            1   

   Mother's qualification  Father's qualification  Mother's occupation  ...  \
0                      13                      10                    6  ...   
1                       1                       3                    4  ...   
2                      22                      27                   10  ...   
3                      23                      27                    6  ...   
4                      22                      28                   10  ...   

   Curricular units 2nd sem (credited)  Curricular units 2nd sem (enrolled)  \
0                                    0                                    0   
1                                    0                                    6   
2                                    0                                    6   
3                                    0                                    6   
4                                    0                                    6   

   Curricular units 2nd sem (evaluations)  \
0                                       0   
1                                       6   
2                                       0   
3                                      10   
4                                       6   

   Curricular units 2nd sem (approved)  Curricular units 2nd sem (grade)  \
0                                    0                          0.000000   
1                                    6                         13.666667   
2                                    0                          0.000000   
3                                    5                         12.400000   
4                                    6                         13.000000   

   Curricular units 2nd sem (without evaluations)  Unemployment rate  \
0                                               0               10.8   
1                                               0               13.9   
2                                               0               10.8   
3                                               0                9.4   
4                                               0               13.9   

   Inflation rate   GDP    Target  
0             1.4  1.74   Dropout  
1            -0.3  0.79  Graduate  
2             1.4  1.74   Dropout  
3            -0.8 -3.12  Graduate  
4            -0.3  0.79  Graduate  

[5 rows x 35 columns]

3. Feature elimination (no consistent bottom ranking features found, conclusion all are relevant.)

In [31]:
#https://www.youtube.com/watch?v=vxdVKbAv6as

#Recursive Feature Elimination

from sklearn.feature_selection import RFE
from sklearn.tree import DecisionTreeClassifier #have to use some estimator
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import Perceptron
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import GradientBoostingClassifier

#input split
x=df2.drop(columns=["Target"], axis=1)
y=df2["Target"]

#estimators
l=[DecisionTreeClassifier(), LogisticRegression(), Perceptron(), RandomForestClassifier(), GradientBoostingClassifier()]

for i in range(0,5):
    rfe=RFE(estimator=l[i], n_features_to_select=29)
    rfe.fit(x,y)
    columns_to_delete=[]
    for i, col in zip(range(x.shape[1]), x.columns):
        # print(str(rfe.support_[i]))
        if (str(rfe.support_[i])) == str(False):
            columns_to_delete.append(col)
        # print("{} selected={} ranks={}".format(col, rfe.support_[i], rfe.ranking_[i]))
    print(sorted(columns_to_delete))

# df3=df2.drop(columns=columns_to_delete)
# df3.head()

'''
No consistent negligent features found...
'''

['Curricular units 2nd sem (without evaluations)', 'Daytime/evening attendance', 'Educational special needs', 'International', 'Marital status']


c:\Users\orlan\Desktop\Python ENV\lib\site-packages\sklearn\linear_model\_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,
c:\Users\orlan\Desktop\Python ENV\lib\site-packages\sklearn\linear_model\_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model

['Curricular units 1st sem (evaluations)', "Father's occupation", "Father's qualification", 'Inflation rate', "Mother's qualification"]
['Application mode', 'Curricular units 2nd sem (evaluations)', "Father's occupation", 'GDP', "Mother's qualification"]
['Daytime/evening attendance', 'Educational special needs', 'International', 'Marital status', 'Nacionality']
['Curricular units 2nd sem (without evaluations)', 'Daytime/evening attendance', 'Educational special needs', 'International', 'Previous qualification']


'\nNo consistent negligent features found...\n'

4. Undersample due to class imbalance.

In [60]:
#Undersample to prevent errors due to class imbalance
#Graduate: 2209/3630
#Dropout: 1421/3630
#Delete (2209-1421=788) Graduate data
import random
graduate_df=df2[df2.Target != "Dropout"]
indexes=list(graduate_df.index.values)
# print(len(indexes))
indexes_to_delete=[]

for i in range(788):
    l=len(indexes)
    n=random.randint(0,l-1)
    indexes_to_delete.append(indexes.pop(n))
# print(len(indexes_to_delete))
# print(indexes_to_delete)
df3=df2.drop(indexes_to_delete, axis='index')
df3 = df3.reset_index(drop=True)

graduate_df=df3[df3.Target != "Dropout"]
dropout_df=df3[df3.Target != "Graduate"]

graduate_row=graduate_df.shape[0]
dropout_row=dropout_df.shape[0]
total_rows=df3.shape[0]
print('Graduate Data: {}/{}, Dropout Data: {}/{}.'.format(graduate_row, total_rows, dropout_row, total_rows))
print('Balanced Classes.')


Graduate Data: 1421/2842, Dropout Data: 1421/2842.
Balanced Classes.


5. Change dropout, graduate to Boolean: ready for numpy. (Dropout=0, Graduate=1)

In [61]:
df4=df3.replace({'Target':{'Dropout': 0, 'Graduate': 1}})
df4.head()
df4.to_csv('Data/preprocessed-dataset.csv')

How much data we need?

https://postindustria.com/how-much-data-is-required-for-machine-learning/#:~:text=The%20most%20common%20way%20to,of%20freedom%20a%20model%20has.
https://machinelearningmastery.com/much-training-data-required-machine-learning/

There is a rule of 10. 10 times the amount of features for the size of dataset.
We have 34 features and a dataset of size 2842. With balanced classses.


6. Make test set (20%) and validation-training cross validation sets for hyperparameters.

In [72]:
import numpy as np
data = np.loadtxt(('Data/preprocessed-dataset.csv'), delimiter=',', skiprows=1)
x = data[:,1:35]
y = data[:,35]

print(x)
print(y)


[[ 1.    8.    5.   ... 10.8   1.4   1.74]
 [ 1.    6.    1.   ... 13.9  -0.3   0.79]
 [ 1.    1.    5.   ... 10.8   1.4   1.74]
 ...
 [ 1.    1.    1.   ... 16.2   0.3  -0.92]
 [ 1.    1.    2.   ... 11.1   0.6   2.02]
 [ 1.    1.    1.   ... 13.9  -0.3   0.79]]
[0. 1. 0. ... 1. 0. 0.]


In [74]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

print(x_test, y_test)

#we use sklearn to cross validate, no need to make different splits.

[[ 1.   12.    1.   ... 13.9  -0.3   0.79]
 [ 1.    1.    4.   ... 12.7   3.7  -1.7 ]
 [ 1.    9.    1.   ...  9.4  -0.8  -3.12]
 ...
 [ 1.   12.    1.   ...  7.6   2.6   0.32]
 [ 1.    1.    1.   ... 10.8   1.4   1.74]
 [ 1.    1.    6.   ... 10.8   1.4   1.74]] [1. 1. 0. 0. 0. 1. 1. 0. 1. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 1. 1. 1. 0. 1.
 1. 1. 0. 1. 0. 0. 1. 0. 0. 1. 1. 1. 0. 1. 0. 1. 0. 0. 0. 1. 1. 0. 1. 1.
 1. 0. 1. 1. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 0. 1. 0. 0. 0. 1. 0. 0.
 1. 0. 0. 0. 1. 1. 0. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 1. 0. 0. 1. 0. 0. 0.
 1. 1. 1. 1. 1. 0. 1. 0. 0. 0. 1. 1. 1. 0. 0. 1. 0. 1. 0. 0. 1. 0. 1. 1.
 0. 0. 1. 1. 0. 0. 1. 1. 0. 0. 0. 1. 1. 0. 1. 0. 1. 1. 1. 0. 0. 1. 0. 1.
 1. 1. 0. 0. 0. 1. 0. 0. 0. 0. 1. 1. 1. 1. 0. 0. 0. 1. 1. 1. 0. 0. 1. 0.
 0. 0. 1. 1. 1. 1. 1. 0. 0. 0. 1. 0. 1. 1. 0. 1. 0. 0. 1. 1. 0. 0. 0. 1.
 1. 0. 0. 0. 1. 1. 0. 0. 0. 1. 1. 1. 1. 0. 1. 0. 1. 1. 0. 0. 1. 1. 0. 0.
 0. 1. 1. 1. 0. 1. 1. 1. 1. 0. 1. 1. 0. 0. 1. 1. 1. 0. 0. 1. 1. 0. 1. 1.
 1. 0.